Add delete by tag feature

joshtyf opened the pull request:
Implemented delete by tag functionality
DG and UG updates have not been committed yet

Update: DG and UG updates have been committed

joshtyf committed:
Make DeleteCommandParser parse tag prefix

joshtyf committed:
Add flashcard predicates to DeleteCommand
DeleteCommand will now also contain tag predicates so that it knows which flashcards to delete.

joshtyf committed:
Make parse method get index from argMultimap preamble
argMultimap now contains the index information in the preamble, hence the index should be parsed from the preamble.

joshtyf committed:
Add delete by tag execution logic

joshtyf committed:
Update MESSAGE_USAGE string

joshtyf committed:
Remove unused imports

joshtyf committed:
Fix missing newline in MESSAGE_USAGE

joshtyf committed:
Add delete with tags message

joshtyf committed:
Add the set of tags to be matched to DeleteCommand
DeleteCommand needs a way to access the tags so as to display it for the CommandResult.

joshtyf committed:
Remove TODO

joshtyf committed:
Fix extra whitespace in MESSAGE_USAGE

joshtyf committed:
Start using static method withIndex for DeleteCommand

joshtyf committed:
Add new typical flashcards
RANDOM8 and RANDOM9 are flashcards with the test tag

joshtyf committed:
Make createDeleteWithTagsMessage method public static
This enables the DeleteCommandTest class to be able to create the delete message as well.

joshtyf committed:
Add execute_validTag_sucess test

joshtyf committed:
Add TypicalTags.java

joshtyf committed:
Add additional assert for execute_validTag_success test

joshtyf committed:
Add execute_InvalidTag_success test

joshtyf committed:
Add an InvalidTag

joshtyf committed:
Reformat to change the ordering of methods

joshtyf committed:
Rename existing test methods
With the addition of tag prefixes, the test methods need to be changed to reflect the different args that we are testing against.

joshtyf committed:
Add additional conditions to the overridden equals method
With the inclusion of predicates and tags, there are now different conditions to check for equality.

joshtyf committed:
Add test parse_validTag_returnsDeleteCommand

joshtyf committed:
Add test parse_emptyTag_throwsParseException

joshtyf committed:
Add test parse_validIndexAndTag_throwsParseException

joshtyf committed:
Add test parse_emptyArgs_throwsParseException

joshtyf committed:
Add test parse_invalidPrefix_throwsParseException

joshtyf committed:
Change createDeleteWithTagsMessage to be a private method

joshtyf committed:
Remove usages of createDeleteWithTagsMessages
DeleteCommand.createDeleteWithTagsMessage method is no longer public static, hence we need to manually input the expected message.

josiahkhoo commented on src/main/java/quickcache/logic/commands/DeleteCommand.java:
```diff
@@ -18,36 +22,98 @@
+        return new DeleteCommand(predicate, tagsToMatch);
+    }
+
+    private String createDeleteWithTagsMessage() {
+        StringBuilder sb = new StringBuilder("with tags ");
+        assert isDeleteByTag && tagsToMatch != null;
+        tagsToMatch.forEach(tag -> sb.append(tag.toString()).append(" "));
+        return sb.toString().trim();
```
Perhaps you can make use of `.reduce()` to make this method even more concise.
```java
return tagsToMatch.stream().reduce(("with tags ", (curr, next) -> curr + next.toString() + " ").trim()
```

joshtyf replied:
truly a CS2030 TA, thanks for the suggestion!

josiahkhoo reviewed (approved):
LGTM, just needs to clear the checkstyle as well on top of the docs addition.

joshtyf committed:
Update UserGuide.md
UserGuide is now updated to show instructions for deleting by tag.

joshtyf committed:
Update DeveloperGuide.md
DeveloperGuide is now updated with new implementation of delete command.

joshtyf committed:
Fix checkstyle errors

joshtyf committed:
Change method body to use stream API
Makes the code a bit more concise as suggested by Josiah

joshtyf committed:
Fix checkstyle errors

joshtyf commented:
DG and UG updates have been committed

joshtyf committed:
Change question for flashcards RANDOM8 and RANDOM9
Accidentally changed the question in previous commit and this failed the LogicManagerTest class tests. It works now

ChenXJ98 commented on src/main/java/quickcache/logic/commands/DeleteCommand.java:
```diff
@@ -64,10 +64,12 @@ public static DeleteCommand withPredicate(FlashcardPredicate predicate, Set<Tag>
+                .reduce("with tags ",
+                        (curr, next) -> curr + next.toString() + " ",
+                        String::concat)
+                .trim();
```
Great implementation!

ChenXJ98 reviewed (approved):
LGTM! 👍
